merkle/src/merkle.rs

```diff
@@ -248,6 +258,241 @@ impl<E: Element> Clone for MmapStore<E> {
     }
 }
 
+/// File-mapping version of `MmapStore` with the added `new_with_path` method
+/// that allows to set its path (otherwise a temporary file is used which is
+/// cleaned up after we drop this structure).
+#[derive(Debug)]
+pub struct DiskMmapStore<E: Element> {
+    // Implementing the `store` with `Arc`/`RwLock` to avoid adding lifetimes
+    // parameters to the struct (which might have larger repercussions in the
+    // definitions of the `MerkleTree` and its consumers. Also used for its
+    // coordination mechanisms, but it's not clearly defined if `MerkleTree`
+    // (and `Store`) should be thread-safe (so they might be removed later).
+    store: Arc<RwLock<Option<MmapMut>>>,
+
+    len: usize,
+    _e: PhantomData<E>,
+    file: File,
+    // We need to save the `File` in case we're creating a `tempfile()`
+    // otherwise it will get cleaned after we return from `new()`.
+
+    // We cache the `store.len()` call to avoid accessing the store when
+    // it's offloaded. Not to be confused with `len`, this saves the total
+    // size of the `store` and the other one keeps track of used `E` slots
+    // in the `DiskMmapStore`.
+    store_size: usize,
+
+    // We save the arguments of `new_with_path` to reconstruct it and reload
+    // the `MmapMut` after offload has been called.
+    path: String,
+    // FIXME: Using `String` instead of `&Path` to avoid using lifetimes
+    //  or errors like "the size for values of type `[u8]` cannot be known
+    //  at compilation time" (with `Option<Box<Path>>` or similar).
+    size: Option<usize>,
+    // FIXME: Needed to reconstruct the `new_with_path` call, especially
+    //  since we are adjusting the size there. Ideally this value should
+    //  be deduced from the MT settings and not stored here.
+}
+
+impl<E: Element> ops::Deref for DiskMmapStore<E> {
+    type Target = [E];
+
+    fn deref(&self) -> &Self::Target {
+        unimplemented!()
+    }
+}
+
+impl<E: Element> Store<E> for DiskMmapStore<E> {
+    #[allow(unsafe_code)]
+    fn new(size: usize) -> Self {
+        let byte_len = E::byte_len() * size;
+        let file: File = tempfile().expect("couldn't create temp file");
```

Comment:
There is a risk of this file suddenly vanishing, given that `tempfile` gets cleaned up by the OS if we close all handles to it, we probably want to switch to sth that we explicitly clean up.

Reply:
Could you expand on that scenario? I'm holding on to the file reference especially to avoid that (but I might be missing something), 

https://github.com/filecoin-project/merkle_light/pull/15/files#diff-51a9ffca05030231847ab34ac3206a3aR275

Reply:
I can explain, I am blind :)